source/solvers/heat_transfer.cc

```diff
@@ -40,6 +43,142 @@ void
+                                                        &this->dof_handler);
+              dh_cell->get_dof_indices(heat_dof_indices);
+
+              const auto pic = solid_ph->particles_in_cell(cell);
+              Assert(pic.begin() == particle, ExcInternalError());
+              for (const auto &p : pic)
+                {
+                  double      temperature = 0;
+                  const auto &ref_q       = p.get_reference_location();
+                  const auto &real_q      = p.get_location();
+                  const auto &JxW         = p.get_properties()[0];
+
+                  for (unsigned int k = 0; k < dofs_per_cell; ++k)
+                    {
+                      // Get the temperature at non-quadrature point (particle
+                      // in fluid)
+                      auto &evaluation_point = this->evaluation_point;
+                      temperature += evaluation_point[heat_dof_indices[k]] *
+                                     this->fe->shape_value(k, ref_q);
+                    }
+                  // Assemble the matrix or the rhs
+                  for (unsigned int i = 0; i < dofs_per_cell; ++i)
+                    {
+                      if (assemble_matrix)
+                        {
+                          for (unsigned int j = 0; j < dofs_per_cell; ++j)
+                            {
+                              local_matrix(i, j) +=
+                                rho_cp * penalty_parameter * beta *
+                                this->fe->shape_value(i, ref_q) *
+                                this->fe->shape_value(j, ref_q) * JxW;
+                            }
+                        }
+                      else
+                        {
+                          // Regular residual
+                          local_rhs(i) +=
+                            -rho_cp * penalty_parameter * beta * temperature *
+                              this->fe->shape_value(i, ref_q) * JxW +
+                            rho_cp * penalty_parameter * beta *
+                              solid_temperature->value(real_q, 0) *
+                              this->fe->shape_value(i, ref_q) * JxW;
+                        }
+                    }
+                }
+              const AffineConstraints<double> &constraints_used =
+                this->zero_constraints;
+              auto &system_rhs = this->system_rhs;
+              constraints_used.distribute_local_to_global(local_matrix,
+                                                          local_rhs,
+                                                          heat_dof_indices,
+                                                          this->system_matrix,
+                                                          system_rhs);
+              particle = pic.end();
+            }
+          this->system_matrix.compress(VectorOperation::add);
+          this->system_rhs.compress(VectorOperation::add);
```

Comment:
This is all fine and dandy, but then why did you need to add the matrix and the rhs to the multiphysics interface?

Reply:
They are no longer used; the original code added the Nitsche restriction in the Nitsche solver directly. But it was very complex for nothing I changed the architecture.